sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,829 @@
+    # grad_i = p_ij (1 + ||y_i - y_j||^2)^-1 (y_i - y_j)
+    # This is equivalent to compute_edge_forces in the authors' code
+    # It just goes over the nearest neighbors instead of all the data points
+    # (unlike the non-nearest neighbors version of `compute_gradient_positive')
+    cdef:
+        int ax
+        long i, j, k
+        long K = neighbors.shape[1]
+        long n = val_P.shape[0]
+        float[3] buff
+        float D, Q, pij
+        float C = 0.0
+        float exponent = (dof + 1.0) / -2.0
+    cdef clock_t t1, t2
+    t1 = clock()
+    for i in range(start, n):
+        for ax in range(n_dimensions):
+            pos_f[i * n_dimensions + ax] = 0.0
+        for k in range(K):
+            j = neighbors[i, k]
+            # we don't need to exclude the i==j case since we've 
+            # already thrown it out from the list of neighbors
+            D = 0.0
+            Q = 0.0
+            pij = val_P[i, j]
+            for ax in range(n_dimensions):
+                buff[ax] = pos_reference[i, ax] - pos_reference[j, ax]
+                D += buff[ax] ** 2.0  
+            Q = (((1.0 + D) / dof) ** exponent)
+            D = pij * Q
+            Q /= sum_Q
+            C += pij * log((pij + EPSILON) / (Q + EPSILON))
+            for ax in range(n_dimensions):
+                pos_f[i * n_dimensions + ax] += D * buff[ax]
+    t2 = clock()
+    dt = ((float) (t2 - t1))
+    if verbose > 10:
+        printf("[t-SNE] Computed error=%1.4f in %1.1e ticks\n", C, dt)
+    return C
+
+
+
+cdef void compute_gradient_negative(float[:,:] val_P, 
+                                    float[:,:] pos_reference,
+                                    float* neg_f,
+                                    Node *root_node,
+                                    float* sum_Q,
+                                    float dof,
+                                    float theta, 
+                                    long start, 
+                                    long stop) nogil:
+    if stop == -1:
+        stop = pos_reference.shape[0] 
+    cdef:
+        int ax
+        long i, j
+        long n = stop - start
+        float* force
+        float* iQ 
+        float* pos
+        float* dist2s
+        long* sizes
+        float* deltas
+        long* l
+        int n_dimensions = root_node.tree.n_dimensions
+        float qijZ, mult
+        long idx, 
+        long dta = 0
+        long dtb = 0
+        clock_t t1, t2, t3
+        float* neg_force
+
+    iQ = <float*> malloc(sizeof(float))
+    force = <float*> malloc(sizeof(float) * n_dimensions)
+    pos = <float*> malloc(sizeof(float) * n_dimensions)
+    dist2s = <float*> malloc(sizeof(float) * n)
+    sizes = <long*> malloc(sizeof(long) * n)
+    deltas = <float*> malloc(sizeof(float) * n * n_dimensions)
+    l = <long*> malloc(sizeof(long))
+    neg_force= <float*> malloc(sizeof(float) * n_dimensions)
+
+    for i in range(start, stop):
```

Comment:
Incidentally, this `for` loop is ripe for parallelization. It is the most expensive part of the gradient calculation. The memory access is opportune as well: the tree is only read from, no writing. And there's only a few small temporary variables in the body of the `for` loop, followed by never-overlapping writes to a single row of a force matrix that can be done in any order. I'm not an expert at multiprocessing in Cython, and my initial attempt with `prange` didn't yield any improvements. If there's an easy way to parallelize this for loop, could be a quick win and an easy way to be faster than the reference implementation :)